### PR TITLE
Restore early tracing

### PR DESCRIPTION
**Pull request by marc-hb**
~2 main commits~ Only one left
- ~add mtrace_printf()~ merged in #4389 
- restore early tracking, "logical" revert of commit eca2089eceef0

Feb 2022 EDIT: I did this before adding `mtrace_printf()`. At the time there was no way to print to the `etrace` ONLY, so having the DMA trace working earlier was much more valuable.

### CODE REVIEW

**Comment by marc-hb** [comment]
BTW I've been using `mtrace_printf()` a lot to try to understand why the DMA trace works in plain SOF but not in Zephyr. It's useful for any early init problem
cc: @keyonjie

**Comment by lgirdwood** [comment]
> BTW I've been using `mtrace_printf()` a lot to try to understand why the DMA trace works in plain SOF but not in Zephyr. It's useful for any early init problem
> cc: @keyonjie

I assuming we would have a flow like

1) Zephyr early boot.
2) Zephyr mailbox trace is initialised. All trace now goes here.
3) SOF init runs and inits DMAC.
4) SOF init registers DMA trace backend. Trace now goes here.

**Comment by marc-hb** [comment]
https://sof-ci.01.org/sofpr/PR4334/build9344/devicetest/?model=CML_RVP_SDW&testcase=check-suspend-resume-5 is the usual thesofproject/linux/issues/2943

In the same run check-sof-logger.sh errors should be fixed by https://github.com/thesofproject/sof-test/pull/707, re-running now.

Everything else is green!

Kept as a draft for now because of recent #4333 regression

**Comment by marc-hb** [comment]
SOFCI TEST

**Review by lgirdwood — APPROVED** [review]

File: src/trace/dma-trace.c
Comment (kv2019i) [inline src/trace/dma-trace.c:179]: This looks a bit weird and something to address before merging.

**Review by kv2019i — COMMENTED** [review]
Looks good otherwise, but the "if (false)" is a bit strange and was not in the the original-now-to-be-reverted patch.

File: src/trace/dma-trace.c
Comment (marc-hb) [reply src/trace/dma-trace.c:179]: I don't like this either but I'm very reluctant to spend a few extra days debugging the probably half-initialized memory allocator that I don't know anything about yet. With this, #4327 and others I've already spent an inordinate amount of time fixing issues in a tracing solution that we want to deprecate eventually. @lgirdwood, @keyonjie any clue about this crash or any name of an expert in this area?

The impact of this issue seems small: a tiny amount of memory wasted in the very unusual case where DMA tracing fails to initialize here = itself a much "bigger problem"! In fact I bet no one would have noticed if I had simply omitted this `rfree()` entirely or even worse: not tested error handling (typical). Now that I narrowed down this crash, I think it's better to have this "strange"/commented out code here to save a lot of time to whoever would try to add it back if I it were missing. It's not like we can use gdb, can we?

Thoughts?

Comment (lgirdwood) [reply src/trace/dma-trace.c:179]: Lets not waste time on fixing this, lest spend our time getting the Zephyr DMA backend working.

Comment (plbossart) [reply src/trace/dma-trace.c:179]: I fundamentally disagree on the decision-making process @lgirdwood. Both memory allocation and trace issues are now being pushed under the rug with a let's use Zephyr argument, that's a shortcut that needs a wider discussion.

The direction has to be one of 

a) we have no plans to have Zephyr used in public and Chrome releases on APL..TGL, so we need to fix the trace and memory allocator
b) we formally decide at the TSG that we ditch the existing SOF with XTOS baseline to use Zephyr only.

I am not against Zephyr, far from it, but we need a formal plan for a transition.

Comment (lgirdwood) [reply src/trace/dma-trace.c:179]: Direction is b. Makes no sense maintaining two RTOSes. The trace/memory issue are not critical yet and we have also been living with them for a while.
I will schedule TSC for immediately after v1.8 to discuss transition plan.

**Review by paulstelian97 — APPROVED** [review]
Conditional approve -- in my view an improved comment stating the observed behaviour and why it is fine to ignore in this specific situation would do it justice.

**Comment by marc-hb** [comment]
BYT and CHT were crashing on QEMU, see https://sof-ci.01.org/sofpr/PR4334/build9353/boottest/ and https://github.com/thesofproject/sof/pull/4334/checks?check_run_id=2800699973

This was because `platform_timer_get(NULL)` was being invoked in `vatrace_log()`. Fixed with new commit and new function `platform_safe_get_time()`

Funny how everything else was running OK.

**Comment by marc-hb** [comment]
Test fixes https://github.com/thesofproject/sof-test/pull/710 need to go first.

**Comment by lgirdwood** [comment]
@marc-hb ping when dependencies are met.

**Comment by marc-hb** [comment]
I resubmitted the first 85% of this in #4389 and unlike here all tests are green. Debugging the BDW and BYT failures in https://sof-ci.01.org/sofpr/PR4334/build9425/devicetest  is not high priority right now.

**Comment by lgirdwood** [comment]
> I resubmitted the first 85% of this in #4389 and unlike here all tests are green. Debugging the BDW and BYT failures in https://sof-ci.01.org/sofpr/PR4334/build9425/devicetest is not high priority right now.

Whats the plan for this now ?

**Comment by marc-hb** [comment]
I rebased and resubmitted. Just like before, BDW and BYT are the two platforms still failing:

https://sof-ci.01.org/sofpr/PR4334/build9533/devicetest/?model=BYT_MB_NOCODEC&testcase=check-sof-logger

The DMA trace is empty and the mailbox has this and no banner:

```
 TIMESTAMP      (us)           DELTA  C# COMPONENT          LOCATION                      CONTENT	ktime=214.698s  @  2021-07-01 06:16:09 +0000 UTC

Found valid LDC address after skipping 32 bytes (one line uses 20 + 0 to 16 bytes)
[   208032082.515] (           0.000) c0 dma-trace             src/trace/dma-trace.c:553  ERROR dtrace_add_event(): number of dropped logs = 1
Skipped 824 bytes after the last statement.
```

On BDW both traces are empty.

Interestingly, BYT and BDW are the same two platforms that reproduce bug #4333 every single time (since test determinism fix https://github.com/thesofproject/sof-test/pull/710)

**Comment by marc-hb** [comment]
SOFCI TEST

**Comment by marc-hb** [comment]
Months and multiple DMA changes later, https://sof-ci.01.org/sofpr/PR4334/build11360/devicetest/ shows BDW and BYT still failing the same, in a "stuck DMA" #4333 way.

**Comment by lgirdwood** [comment]
> Months and multiple DMA changes later, https://sof-ci.01.org/sofpr/PR4334/build11360/devicetest/ shows BDW and BYT still failing the same, in a "stuck DMA" #4333 way. Everything else passes.

Sorry, do you mean we have the stuck DMA with or without this PR ?

**Comment by lgirdwood** [comment]
@marc-hb any update here ?

**Comment by marc-hb** [comment]
Still blocked by mysterious trace corruption. Who knows, maybe #5120 will help?

**Comment by marc-hb** [comment]
> Still blocked by mysterious trace corruption. Who knows, maybe #5120 will help?

#5120 was fixed by DMA alignment fix #5329 but it did not seem to help. I rebased this and BDW is still missing its etrace  in https://sof-ci.01.org/sofpr/PR4334/build12063/devicetest

BYT was not available, some reservation problem. Will re-run after it's fixed.

In fact #5329 alignment made this PR more complicated because I had to hardcode the alignment value (see 2ae33281fc595de7140ec6d21cdab9ab79d16df1) because of a new (circular?) DMAC dependency. @jyri any idea here?


Other unrelated errors spotted in https://sof-ci.01.org/sofpr/PR4334/build12063/devicetest
- #5352 in https://sof-ci.01.org/sofpr/PR4334/build12063/devicetest/?model=WHL_UPEXT_HDA_ZEPHYR&testcase=test-speaker
- https://github.com/thesofproject/linux/issues/3012 in https://sof-ci.01.org/sofpr/PR4334/build12063/devicetest/?model=CML_RVP_SDW&testcase=check-suspend-resume-with-playback-5

**Comment by marc-hb** [comment]
SOFCI TEST

**Comment by ujfalusi** [comment]
> > Still blocked by mysterious trace corruption. Who knows, maybe #5120 will help?
> 
> #5120 was fixed by DMA alignment fix #5329 but it did not seem to help. I rebased this and BDW is still missing its etrace in https://sof-ci.01.org/sofpr/PR4334/build12063/devicetest

etrace has nothing to do with dma-trace or am I missing something?

It could be possible that the dtarce init fails earlier than we would print the marker to the mtrace window (which is the etrace) and since the sof-logger is looking for the banner it is not printing out anything?

It would be interesting to see what is in there when the banner can not be found... It might just have the right information and we might want to add the SHM banner when the mtrace is initialized, which happens before the dtrace?

**Comment by marc-hb** [comment]
> It would be interesting to see what is in there when the banner can not be found... It might just have the right information and we might want to add the SHM banner when the mtrace is initialized, which happens before the dtrace?

The etrace is completely empty. Did you look at the test logs? I think they answer several of your questions.

**Comment by marc-hb** [comment]
I filed new "empty etrace" bug in
- #5385
I should have filed this a long time ago sorry.

This draft PR always reproduces an empty etrace on BDW and BYT but it can also happen WITHOUT this PR.

File: src/trace/dma-trace.c
Comment (marc-hb) [reply src/trace/dma-trace.c:280]: This is a hack, I merely copied the value from the disabled code above.

**Comment by kv2019i** [comment]
Stale PR

**Pull request closed without merging**
